### ISSUE TITLE
Feature request: nix search should provide some caching functionalities

### ISSUE THREAD

**Issue by tobiasBora** [NixOS/nix#6421]
**Describe the bug**

When doing `nix search nixpkgs firefox` (flake version), it takes a lot of time (several minutes) as it needs to download the latest nixpkgs and build a search cache. The problem is that this cache is invalidated multiple times per day.

One workaround is to pin nixpkgs, but then this can produce even more nasty bugs: I realized once that I forgot that nixpkgs was pinned 2 years ago, explaining why I was installing outdated softwares. It would be better instead to be able to specify a duration for the cache (like 1 month), with methods to automatically update the cache when it gets too old, or when the users forces the cache to be refreshed. A warning could be displayed with the age of the cache.

**Steps To Reproduce**

1. Run `nix search nixpkgs firefox`, maybe two times at an interval of one day

**Expected behavior**

I expect `search` to be really quick.

**`nix-env --version` output**
nix-env (Nix) 2.5pre20211007_844dd90

**Comment by vcunat** [comment]
Tip: in case you use channels, you [can do](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/programs/command-not-found/command-not-found.nix) stuff like
```
$ bar
The program 'bar' is not in your PATH. You can make it available in an
ephemeral shell by typing:
  nix-shell -p bar
```
It's very cheap and fast.

**Comment by tobiasBora** [comment]
True, but it only helps when you know the name of the executable, when you are not searching for libraries, and I guess when the channels are up to date.

**Comment by vcunat** [comment]
These mappings tend to change rarely; you'll usually be fine even with years old channel.

**Comment by tobiasBora** [comment]
The problem is that pinning will break the rest of the worflow: people (at least I) will forget that the channel is pinned, and then will start installing old softwares with incompatible glibc…

**Comment by tobiasBora** [comment]
I mean, if I do `nix run nixpkgs#firefox`, it will use the same pinned channel as `nix search nixpkgs`. And if it's outdated, then I'll install outdated packages without even noticing…

**Comment by tobiasBora** [comment]
But what's so complicated about caching ? I feel it should be nothing more than a small condition

**Comment by vcunat** [comment]
I'm not saying anything against it.  Just offering tips, as it seems very unlikely to be done by someone... like this month at least ;-)  (just my guess, though)

**Comment by thufschmitt** [comment]
> It would be better instead to be able to specify a duration for the cache (like 1 month), with methods to automatically update the cache when it gets too old, or when the users forces the cache to be refreshed.

That’s not very fined-grained, but the [tarball-ttl](https://nixos.org/manual/nix/stable/command-ref/conf-file.html#conf-tarball-ttl) option should do just that. Maybe it would be interesting to be able to specify it on a per-flake level, like `github:nixos/nixpkgs?cache-ttl=1d`

**Comment by tobiasBora** [comment]
@vcunat Oh ok, thanks, I though you were saying that caching is useless. 

Thanks for pointing to this, but there is something I'm not sure to understand: the default value for this parameter is 4294967295 seconds, which is more than a century. But on my system (I don't think I even modified this parameter), I get cache invalidated way before (maybe less than one day).

**Comment by thufschmitt** [comment]
> the default value for this parameter is 4294967295 seconds, which is more than a century

Mh, that seems to be a bug in the doc generation, the [actual default](https://github.com/NixOS/nix/blob/master/src/libstore/globals.hh#L545) should be one hour

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
